### PR TITLE
Documentation on how to build navigation

### PR DESCRIPTION
**Pull request by tomjohnson1492**
I added a documentation page on how to build navigation for your site. This topic is primarily intended for users who have a lot of pages on their site (such as for documentation websites), and want to build a more robust sidebar navigation.

Jekyll combines Liquid with YAML in interesting ways that aren't really documented clearly in the existing docs, except for a brief reference [here](http://jekyllrb.com/docs/datafiles/#the-data-folder). You can read about Liquid on Shopify and YAML in YAML's docs, but exactly how you store YAML files in a Jekyll project and iterate through them using Liquid loops and filters to generate lists of pages is something that isn't clear to a lot of people. (You can see origins of these questions in [previous help issues](https://github.com/jekyll/jekyll-help/issues/266).)  The documentation on navigation would fit well into the Jekyll docs.

@jekyll/documentation

### CODE REVIEW

**Comment by ashmaroli** [comment]
@DirtyF How about hosting **all `"how to.."`** stuff on another document entirely? (or a subdomain e.g. `tips.jekyllrb.com`?)

/cc @parkr

**Comment by tomjohnson1492** [comment]
Are you proposing that *all* documentation be moved into a subdomain? Or just some types of documentation? 

If it's just *some* types of documentation, my suggestion is that rather than fragmenting the documentation into different domains, we instead create an Advanced page that simply lists the advanced tutorials. 

I get that we probably don't want Jekyll's documentation to seem too complicated. (All product managers want their product to seem simple.) But the result is often a paradox. The more "simple" we try to make something by removing documentation, the harder it can become for people to use the product. As soon as people get passed the first step of installing Jekyll, they're confused at how to go to the next step.

The solution to providing simple instructions for novice users and advanced instructions for advanced users is to separate the info out a bit. Put the advanced content into an advanced area. This is why I proposed adding a link to the navigation document from http://jekyllrb.com/docs/datafiles/. The datafiles doc starts out easy, and when you want to do something more, you transition to more advanced info (progressive disclosure).

But maybe I'm misinterpreting the proposal here. Maybe the subdomain idea is to facilitate something else altogether, such as simplifying merge updates? Would decoupling docs from the other Jekyll code make it easier to deploy updates?

Re tips in the subdomain, there's already http://jekyll.tips/, so that wouldn't work. (Actually, I'd love to see Mike's info be merged with the existing Jekyll doc. I don't know why that's separate in the first place. His tutorials and info are golden.)

The most logical alternative for a subdomain is docs.jekyllrb.com.

**Comment by ashmaroli** [comment]
What I'm proposing is that **all** documentation related to non-basic customization of Jekyll be moved to a separate document or subdomain ( new repo ) by itself. `http://tips.jekyllrb.com` was only a suggestion. In theory it will work because `tips` will be a subdomain of `jekyllrb.com`, no way related to `http://jekyll.tips`. But should the idea be considered, we could think of using a different subdomain entirely to avoid confusion with the latter site. 

IMO, `jekyllrb.com` being official documentation should not disseminate tutorials.

**Comment by pathawks** [comment]
> What I'm proposing is that all documentation related to non-basic customization of Jekyll be moved to a separate document

I disagree that making it more difficult to find or contribute to documentation would solve any problem.

**Comment by tomjohnson1492** [comment]
I think it's a good idea to have different areas of documentation. I can see value in having a section designated for tutorials, another section designated for reference, another for getting started. And I can see how this topic on implementing navigation would fit into a tutorials section.

But as far as shifting around the different types of content onto different subdomains, I'm not a fan of that. It will make linking more difficult, as well as potentially create confusion among users. 

How about just adding a link to "Tutorials" in the sidebar, and then on that page, add a link to "Implementing Navigation." Or create a _tutorials collection and put them there. Then they could be under jekyllrb.com/tutorials instead of jekyllrb.com/docs. I like that idea.

I have a few other tutorials brewing in the back of my mind. For example, implementing media RSS feeds, converting an HTML theme into a jekyll site, implementing translation, validating links, and more. All of these topics could comfortably fit into a Tutorials collection.

Really, users don't care as long as they can find the information and it answers their question. Good documentation has a mix of reference, tutorials, how-to topics, and more. Think about it -- the foundation of all developer documentation is a "Hello World" tutorial.

**Comment by ashmaroli** [comment]
:+1:  to more tutorials.. I would just like them all collected under a single dedicated document/sub-domain.
:+1:  to _tutorials collection folder. It addresses my point with minimal interference.

**Comment by tomjohnson1492** [comment]
Do you want me to create the collection for this? I can. I just wasn't sure if you were waiting for me to do that.

File: docs/_docs/navigation.md [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+  url: deployment.html
+```
+
+**Liquid**
+
+```liquid
+{% raw %}<h2>{{ site.data.samplelist.docs_list_title }}</h2>
+<ul>
+   {% for item in site.data.samplelist.docs %}
+   <li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: contents of liquid blocks are indented as well..
```
<ul>
  {% for item in site.data.samplelist.docs %}
    <li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
  {% endfor %}
</ul>
```

File: docs/_docs/navigation.md [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+
+**Result**
+<div class="result">
+<h2>ACME Documentation</h2>
+<ul>
+<li><a href="introduction.html" alt="Introduction">Introduction</a></li>
+<li><a href="configuration.html" alt="Configuration">Configuration</a></li>
+<li><a href="deployment.html" alt="Deployment">Deployment</a></li>
+</ul>
+</div>
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: indentation required..

File: docs/_docs/navigation.md [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+Suppose you wanted to sort the list by the `title`. To do this, convert the reference to the `docs` collection to a variable, and then apply Liquid's `sort` filter to the variable:
+
+**Liquid**
+
+```liquid
+{% raw %}{% assign doclist = site.data.samplelist.docs | sort: 'title'  %}
+{% for item in doclist %}
+<li><a href="{{ item.url }}" alt="{{ item.title }}">{{ item.title }}</a></li>
+{% endfor %}{% endraw %}
+```
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: indentation required..

File: docs/_docs/navigation.md [hunk 4]
```diff
@@ -0,0 +1,570 @@
+---
+layout: docs
+title: Navigation
+permalink: /docs/navigation
+---
+
+<style>
```
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+.result {
+border: 1px solid yellow;
```
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md (hunk 1 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md (hunk 2 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md (hunk 3 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

**Comment by tomjohnson1492** [comment]
I made the necessary updates around indentation. Submitting for re-review.

File: docs/_docs/navigation.md [hunk 6]
```diff
@@ -319,19 +323,19 @@ sidebar: toc
 
 ```liquid
 {% raw %}<ul>
-{% for item in site.data.samplelist[page.sidebar] %}
-<li><a href="{{ item.url }}">{{ item.title }}</a></li>
-{% endfor %}
+    {% for item in site.data.samplelist[page.sidebar] %}
+    <li><a href="{{ item.url }}">{{ item.title }}</a></li>
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: requires additional indentation here.. :wink:

File: docs/_docs/navigation.md [hunk 7]
```diff
@@ -353,8 +357,8 @@ In addition to inserting items from the YAML data file into your list, you also
 
 ```liquid
 {% raw %}{% for item in site.data.samplelist.docs %}
-<li class="{% if item.url == page.url %}active{% endif %}">
-<a href="{{ item.url }}">{{ item.title }}</a></li>
+    <li class="{% if item.url == page.url %}active{% endif %}">
+    <a href="{{ item.url }}">{{ item.title }}</a></li>
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: - the anchor tag is nested.. so would need to be indented..
- the closing list-item tag on a new-line..

File: docs/_docs/navigation.md [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+```liquid
+{% raw %}{% for item in site.data.samplelist.docs %}
+    <li class="{% if item.url == page.url %}active{% endif %}">
+    <a href="{{ item.url }}">{{ item.title }}</a></li>
+{% endfor %}{% endraw %}
+```
+
+**Result**
+
+<style>
```
Comment (ashmaroli) [inline docs/_docs/navigation.md]: `<style>` => External CSS

File: docs/_sass/_style.scss [hunk 9]
```diff
@@ -1031,3 +1031,11 @@ code.output {
   clip: rect(0, 0, 0, 0);
   border: 0;
 }
+
+.result {
+  border: 1px solid yellow;
```
Comment (DirtyF) [inline docs/_sass/_style.scss]: I'd rather simply apply `highlight` for now. We can discuss style in a separate PR.

result-styles

File: docs/_docs/navigation.md [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+        {% endif %}
+    {% endfor %}
+</ul>{% endraw %}
+```
+
+**Result**
+
+<div class="result">
+   <ul>
+      <li><a href="introduction.html">Introduction</a></li>
```
Comment (DirtyF) [inline docs/_docs/navigation.md]: All these clickable links will generate 404 errors 😭 

Shouldn't we use `#` in href for all examples here?

File: docs/_sass/_style.scss (hunk 9 above)
Comment (tomjohnson1492) [inline docs/_sass/_style.scss]: Thanks for the comments. I'll make the updates in a couple of days (I'm finishing up some presentations).

**Comment by ashmaroli** [comment]
Some more ideas:
  - Collect all tutorials under a new **category:** `tutorials`
  - Like how our release news-items are published, allow author to have their name displayed. (optional: link to their blog/website)

/cc @DirtyF, @pathawks

File: docs/_docs/navigation.md (hunk 6 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md (hunk 7 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md (hunk 10 above)
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_docs/navigation.md [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+Here's the code for getting lists of pages grouped under their corresponding category headers:
+
+**Liquid**
+
+```liquid
+{% raw %}{% assign mydocs = site.docs | group_by: 'category' %}
+{% for cat in mydocs %}
+<h2>{{ cat.name | capitalize }}</h2>
+    <ul>
+        {% assign items = cat.items | sort: 'weight' %}
```
Comment (tomjohnson1492) [inline docs/_docs/navigation.md]: updated

File: docs/_sass/_style.scss (hunk 9 above)
Comment (tomjohnson1492) [inline docs/_sass/_style.scss]: updated, but i did leave the padding style b/c without it, the text is right against the line.

**Comment by tomjohnson1492** [comment]
I made the updates as noted in the reviews. Submitting for review to merge. Thanks, and sorry for my delay in getting to this.

**Comment by tomjohnson1492** [comment]
@ashmaroli We talked about making tutorials a collection and putting them into another section. I'm guessing I should make that change, right? This will involve me editing the config file and adding more modifications in the nav somewhere.

**Comment by ashmaroli** [comment]
I would like that. But I'll leave the making-the-executive-decisions to @DirtyF  :smiley:

**Comment by parkr** [comment]
> making tutorials a collection and putting them into another section. I'm guessing I should make that change

@tomjohnson1492 If you're up for it, yes! You can use a lot of the existing templates for `docs`, but change `site.docs` to `site.tutorials` in all the loops and `where` clauses and add `tutorials` to to the config. It may force us to re-think our top-level navigation ("HOME DOCS HELP GITHUB" etc) but for now, I think let's go for it!

**Comment by tomjohnson1492** [comment]
I made an update here that you'll want to look at. Following @parkr's  green-light, I added a new collection called **tutorials** and made it function similar to **docs**. I also added an overview page to the Tutorials section (in addition to the navigation topic that already exists). I think this section will help create a space for a wider number of community contributions -- I can already think of about 5 tutorials I want to add here.

@jekyll/documentation

File: docs/_docs/datafiles.md [hunk 12]
```diff
@@ -71,6 +71,9 @@ You can now render the list of members in a template:
 {% endraw %}
 ```
 
+{: .note .info }
+If your Jekyll site has a lot of pages, such as with documentation websites, see the detailed examples in [how to build robust navigation for your site](../navigation).
```
Comment (DirtyF) [inline docs/_docs/datafiles.md]: We need to update the link to tutorials here.

**Comment by DirtyF** [comment]
@tomjohnson1492  Shouldn't we point to the data documentation reference somewhere in the navigation tutorial?

File: docs/_docs/datafiles.md (hunk 12 above)
Comment (tomjohnson1492) [inline docs/_docs/datafiles.md]: ahh, good catch. updated.

**Comment by tomjohnson1492** [comment]
I added a reference to the data files topic. Good catch there.

**Comment by tomjohnson1492** [comment]
Just wondering, do you happen to know when this PR will be merged? Does the merging only happen during release windows, or are merges done continuously? My other PR tutorials are dependent on this one.

File: docs/_includes/primary-nav-items.html [hunk 13]
```diff
@@ -5,6 +5,9 @@
   <li class="{% if page.url contains '/docs/' %}current{% endif %}">
     <a href="/docs/home/">Docs</a>
   </li>
+  <li class="{% if page.url contains '/tutorials/' %}current{% endif %}">
+    <a href="/tutorials/home/">Tutorials</a>
+  </li>
```
Comment (tomjohnson1492) [inline docs/_includes/primary-nav-items.html]: I could remove the "Tutorials" link from the homepage nav and list it on the ["Help" page](http://jekyllrb.com/help/) right under Documentation. Would that work?

Comment (parkr) [inline docs/_includes/primary-nav-items.html]: Yeah, that would work for now 😄

**Comment by tomjohnson1492** [comment]
@parkr I made updates from your review. I moved the "Tutorials" link to the help page and removed the yellow border around the result content.

I'd really like to get this PR merged so that I can add other tutorials here as well. Let me know if I need to do anything else to speed this along. Thanks.

**Comment by parkr** [comment]
Let's do it! Thanks 😄 

@jekyllbot: merge +site

**Comment by parkr** [comment]
> We are investigating issues serving web hooks as well as assets on GitHub.com.
> – [GitHub Status](https://status.github.com/messages)

@jekyllbot uses webhooks, WOMP

**Comment by tomjohnson1492** [comment]
wow, that was fast. Thanks!

**Pull request merged**